### PR TITLE
Add new `RSpec/RedundantPredicateMatcher` cop

### PR DESCRIPTION
**Pull request by ydah**
Fix: https://github.com/rubocop/rubocop-rspec/issues/1689

______________________________________________________________________

Before submitting the PR make sure the following are checked:

- [x] Feature branch is up-to-date with `master` (if not - rebase it).
- [x] Squashed related commits together.
- [x] Added tests.
- [x] Updated documentation.
- [x] Added an entry to the `CHANGELOG.md` if the new code introduces user-observable changes.
- [x] The build (`bundle exec rake`) passes (be sure to run this locally, since it may produce updated documentation that you will need to commit).

If you have created a new cop:

- [x] Added the new cop to `config/default.yml`.
- [x] The cop is configured as `Enabled: pending` in `config/default.yml`.
- [x] The cop is configured as `Enabled: true` in `.rubocop.yml`.
- [x] The cop documents examples of good and bad code.
- [x] The tests assert both that bad code is reported and that good code is not reported.
- [x] Set `VersionAdded: "<<next>>"` in `default/config.yml`.

### CODE REVIEW

File: spec/rubocop/cop/rspec/builtin_matcher_spec.rb [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+
+RSpec.describe RuboCop::Cop::RSpec::BuiltinMatcher do
+  it 'registers an offense when using `be_all`' do
+    expect_offense(<<~RUBY)
+      expect(foo).to be_all(bar)
+                     ^^^^^^^^^^^ Use `all` instead of `be_all`.
+    RUBY
+
+    expect_correction(<<~RUBY)
+      expect(foo).to all(bar)
```
Comment (pirj) [inline spec/rubocop/cop/rspec/builtin_matcher_spec.rb:11]: Is this correct?

File: spec/rubocop/cop/rspec/builtin_matcher_spec.rb [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+  end
+
+  it 'registers an offense when using `be_exists`' do
+    expect_offense(<<~RUBY)
+      expect(foo).to be_exists(bar)
+                     ^^^^^^^^^^^^^^ Use `exist` instead of `be_exists`.
+    RUBY
+
+    expect_correction(<<~RUBY)
+      expect(foo).to exist(bar)
```
Comment (pirj) [inline spec/rubocop/cop/rspec/builtin_matcher_spec.rb:77]: Does it accept an argument?

File: spec/rubocop/cop/rspec/builtin_matcher_spec.rb [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+    RUBY
+
+    expect_correction(<<~RUBY)
+      expect(foo).to start_with(bar)
+    RUBY
+  end
+
+  it 'does not register an offense when using built-in matcher' do
+    expect_no_offenses(<<~RUBY)
+      expect(foo).to all(bar)
```
Comment (pirj) [inline spec/rubocop/cop/rspec/builtin_matcher_spec.rb:127]: All again

File: lib/rubocop/cop/rspec/builtin_matcher.rb [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+      # @example
+      #   # bad
+      #   expect(foo).to be_exist(bar)
+      #   expect(foo).not_to be_include(bar)
+      #
+      #   # good
+      #   expect(foo).to exist(bar)
+      #   expect(foo).not_to include(bar)
+      #
+      class BuiltinMatcher < Base
```
Comment (pirj) [inline lib/rubocop/cop/rspec/builtin_matcher.rb:17]: There’s a wip cop naming guideline saying yhat the name should indicate the offence. I’d suggest RedundantPredicateMatcherUse

File: spec/rubocop/cop/rspec/builtin_matcher_spec.rb (hunk 1 above)
Comment (ydah) [reply spec/rubocop/cop/rspec/builtin_matcher_spec.rb]: I have changed to the following correct code:
```ruby
expect(foo).to be_all be_odd
```

File: spec/rubocop/cop/rspec/builtin_matcher_spec.rb (hunk 2 above)
Comment (ydah) [reply spec/rubocop/cop/rspec/builtin_matcher_spec.rb]: It seems accept one argument:
https://rubydoc.info/github/rspec/rspec-expectations/RSpec/Matchers#exist-instance_method

```ruby
expect(File).to exist("path/to/file")
```

File: spec/rubocop/cop/rspec/builtin_matcher_spec.rb (hunk 3 above)
Comment (ydah) [reply spec/rubocop/cop/rspec/builtin_matcher_spec.rb]: https://github.com/rubocop/rubocop-rspec/pull/1694#discussion_r1293057164

Ditto.

File: lib/rubocop/cop/rspec/builtin_matcher.rb (hunk 4 above)
Comment (ydah) [reply lib/rubocop/cop/rspec/builtin_matcher.rb]: I see. How about `RedundantPredicateMatcher`?

File: spec/rubocop/cop/rspec/redundant_predicate_matcher_spec.rb [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+  it 'registers an offense when using `be_all`' do
+    expect_offense(<<~RUBY)
+      expect(foo).to be_all be_odd
+                     ^^^^^^^^^^^^^ Use `all` instead of `be_all`.
+    RUBY
+
+    expect_correction(<<~RUBY)
+      expect(foo).to all be_odd
+    RUBY
+  end
```
Comment (pirj) [inline spec/rubocop/cop/rspec/redundant_predicate_matcher_spec.rb:13]: Is there an example with `all`?
```ruby
            expect(units).to be_all { |unit| unit.line_item == line_item_2 }
                             ^^^^^^
```
Would the conversion be correct?

Comment (pirj) [inline spec/rubocop/cop/rspec/redundant_predicate_matcher_spec.rb:13]: Yet another
```ruby
        expect(deployment_refs_exist?).to be_all(false)
                                          ^^^^^^^^^^^^^
```

Comment (pirj) [inline spec/rubocop/cop/rspec/redundant_predicate_matcher_spec.rb:13]: This may deserve its own cop:
```
/Users/pirj/source/real-world-rspec/gitlabhq/spec/services/deployments/archive_in_project_service_spec.rb:29:49: C: [Correctable] RSpec/RedundantPredicateMatcher: Use all instead of be_all.
        expect(deployments.map(&:archived?)).to be_all(false)
                                                ^^^^^^^^^^^^^
```
It could be `expect(deployments).to all be_archived(false)`. Looks bad with such an informal negation, though.

Comment (ydah) [reply spec/rubocop/cop/rspec/redundant_predicate_matcher_spec.rb:13]: Since the auto-correction is not correct, would it be better not to make this an offense. So I change not offense for this case.

Comment (ydah) [reply spec/rubocop/cop/rspec/redundant_predicate_matcher_spec.rb:13]: Ditto. I change to not offense for this case.

Comment (ydah) [reply spec/rubocop/cop/rspec/redundant_predicate_matcher_spec.rb:13]: This looks like a good jurisdiction for other cops.

File: lib/rubocop/cop/rspec/redundant_predicate_matcher.rb [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+        private
+
+        def message(bad_method, good_method)
+          format(MSG, bad: bad_method, good: good_method)
+        end
+
+        def replacable_arguments?(node)
+          if node.method?(:be_all)
+            node.first_argument.send_type?
```
Comment (pirj) [inline lib/rubocop/cop/rspec/redundant_predicate_matcher.rb:45]: Wondering if all send nodes are fine?
```ruby
expect(foo).to be_all be_odd
```
is. But
```ruby
let(:expected_foo_element_value) { 0 }
it do
  expect(foo).to be_all(expected_foo_element_value)
```
? It's also a send node, but can't be used as
```ruby
  expect(foo).to all expected_foo_element_value
```

It not only has to be a send, it also has to be a matcher. But we can't know it for sure in our static code checks if it is.

Comment (ydah) [reply lib/rubocop/cop/rspec/redundant_predicate_matcher.rb:45]: Ah, I see. That's for sure. Thanks for catching this. I wonder if I should exclude `all` from the scope, since it's hard to check if it's a matcher and I think it's likely to generate a lot of false positives. What do you think?

Comment (pirj) [reply lib/rubocop/cop/rspec/redundant_predicate_matcher.rb:45]: Yes, that’s something I’ve been thinking, too. Maybe ‘be_all’ can be ignored for autocorrection when there’s a send node? We already ignore the block form. I don’t think ‘be_all(true)’ eg is problematic.

Comment (ydah) [reply lib/rubocop/cop/rspec/redundant_predicate_matcher.rb:45]: We make it so that it is ignored in the autocorrection. 
Maybe we might as well make it not an offense in the first place.

**Review by pirj — APPROVED** [review]
Fantastic! Thank you!

**Comment by pirj** [comment]
Broke on:
```ruby
      expect(json[:submissions]).to be_all do |ss|
        ss.key?("submission_history") && ss["submission_history"].empty?
      end
```
with:
```
undefined method `send_type?' for nil:NilClass
/Users/pirj/source/rubocop-rspec/lib/rubocop/cop/rspec/redundant_predicate_matcher.rb:47:in `replacable_arguments?'
/Users/pirj/source/rubocop-rspec/lib/rubocop/cop/rspec/redundant_predicate_matcher.rb:27:in `on_send'
```
Wondering why `if node.parent.block_type?` didn't work 🤔

**Comment by pirj** [comment]
Otherwise:
```
57497 files inspected, 883 offenses detected, 742 offenses autocorrectable
```
🎉 

False positive? [`match` takes a mandatory argument](https://github.com/rspec/rspec-expectations/blob/562d327cb951575d13db2a43b232c9ca1456ea6a/lib/rspec/matchers.rb#L697).
```
/Users/pirj/source/real-world-rspec/gitlabhq/spec/lib/gitlab/zoom_link_extractor_spec.rb:30:58: C: [Correctable] RSpec/RedundantPredicateMatcher: Use match instead of be_match.
        expect(described_class.new('issue text')).not_to be_match
                                                         ^^^^^^^^
```
```ruby
module Gitlab
  class ZoomLinkExtractor
    # ...
    def match?
      ZOOM_REGEXP.match?(@text)
    end
```

Notes for future cops:
```ruby
expect(json.count).to be_equal 1
expect(json[0]["group_submissions"]).to be_equal nil
expect(effective_paths.size).to be_eql(1)
```

**Comment by ydah** [comment]
Sorry for the delay. I updated this PR with the following corrections:
https://github.com/rubocop/rubocop-rspec/pull/1694#issuecomment-1745654090
https://github.com/rubocop/rubocop-rspec/pull/1694#issuecomment-1745680669

File: CHANGELOG.md [hunk 7]
```diff
@@ -21,6 +23,7 @@
 - Fix an infinite loop error when `RSpec/ExcessiveDocstringSpacing` finds a description with non-ASCII leading/trailing whitespace. ([@bcgraham])
 - Fix an incorrect autocorrect for `RSpec/ReceiveMessages` when return values declared between stubs. ([@marocchino])
 - Fix a false positive `RSpec/Focus` when chained method call and inside define method. ([@ydah])
+- Split `RSpec/FilePath` into `RSpec/SpecFilePathSuffix` and `RSpec/SpecFilePathFormat`. `RSpec/FilePath` cop is enabled by default, the two new cups are pending and need to be enabled explicitly. ([@ydah])
```
Comment (pirj) [inline CHANGELOG.md:26]: Unrelated?

**Review by pirj — APPROVED** [review]
Thank you!

File: CHANGELOG.md (hunk 7 above)
Comment (ydah) [reply CHANGELOG.md]: Thank you! I updated this PR.
